File: Project/jt_box/utilities/jt_covariance_incremental.py
import numpy as np


def streaming_cov(x, avg_old=None, cov_old=None, m_old=0.0):
    m_old = float(m_old)
    m_obs = float(x.shape[0])
    
    avg_obs = np.mean(x, 0)

    m_new = m_old + m_obs

    if m_old <= 0.0:

        avg_new = avg_obs

        x1 = x - avg_obs

        cov_obs = np.dot(x1.T, x1)

        newweight = m_new - 1.0 # why -1?
        cov_new = cov_obs / newweight

    else:

        avg_new = avg_old + (avg_obs - avg_old) * (m_obs / m_new)

        x1 = x - avg_old
        x2 = x - avg_new

        cov_obs = np.dot(x1.T, x2) # why not x2'x2?

        newweight = m_new - 1.0 # why -1?
        cov_new = cov_old * (m_old - 1.0) / newweight + cov_obs / newweight

    return avg_new, cov_new, m_new

# Create data
m = 360*60*4
n = 100
step = 360/2
data = (10 * np.random.rand(m, n)).astype("float64")

# Normal covariance
normal_cov = np.cov(data, rowvar=False)

# Incremental covariance
update_cov = (None, None, 0.0)
for i in range(0, m, step):
    update_cov = streaming_cov(data[i:i+step, :], *update_cov)

# Normal
print(normal_cov)

# Incremental
print(update_cov[1])

# Difference
print(update_cov[1] - normal_cov)

File: Project/jt_box/utilities/jt_covariance_incremental.m
function [avg_new, cov_new, m_new] = jt_covariance_incremental(obs, avg_old, cov_old, m_old)

m_obs = size(obs, 1);
avg_obs = mean(obs, 1);

% Sample count
m_new = m_obs + m_old;

% First data
if m_old <= 0

    % Average
    avg_new = avg_obs;

    % Zero mean
    x1 = bsxfun(@minus, obs, avg_obs);

    % Covariance
    cov_obs = x1' * x1;
    cov_new = cov_obs / (m_new - 1);

% Update
else

    % Average
    avg_new = avg_old + (avg_obs - avg_old) * (m_obs / m_new);

    % Zero mean
    x1 = bsxfun(@minus, obs, avg_old);
    x2 = bsxfun(@minus, obs, avg_new);

    % Covariance
    cov_obs = x1' * x2;
    cov_new = cov_obs / (m_new - 1) + cov_old * ((m_old - 1) / (m_new - 1));

end

function test_case()

m = 360*4*60;
n = 100;
step = 360/2;
X = double(rand(m, n));

% Incremental
avg_inc = [];
cov_inc = [];
m_inc = 0;
for i = 1:m/step
    obs = X(1 + (i-1) * step:i * step, :);
    [avg_inc, cov_inc, m_inc] = jt_covariance_incremental(obs, avg_inc, cov_inc, m_inc);
end

% Non-incremental
X = bsxfun(@minus, X, mean(X, 1));
cov_noninc = cov(X);

disp(max(abs(cov_inc(:) - cov_noninc(:))));

disp(m_inc-m);